File: inserter/src/main/java/com/quid/inserter/Pay/repository/PaymentRepository.java
package com.quid.inserter.Pay.repository;

import com.quid.inserter.Pay.domain.Payment;
import java.util.List;
import java.util.Optional;
import lombok.extern.slf4j.Slf4j;
import org.springframework.jdbc.core.simple.JdbcClient;
import org.springframework.stereotype.Repository;
import org.springframework.transaction.annotation.Transactional;

@Slf4j
@Repository
public class PaymentRepository {

    private final JdbcClient jdbcClient;

    public PaymentRepository(JdbcClient jdbcClient) {
        this.jdbcClient = jdbcClient;
    }

    @Transactional
    public void save(Payment payment) {
        log.info("== SAVE PAYMENT ==");
        String query = "INSERT INTO Payment (transactionKey, referenceKey, status, amount, currency, regDate, modDate) VALUES (?, ?, ?, ?, ?, ?, ?)";

        jdbcClient.sql(query)
            .params(List.of(payment.transactionKey(), payment.referenceKey(), payment.status().name(), payment.amount(), payment.currency(), payment.regDate(), payment.modDate()))
            .update();
    }

    @Transactional(readOnly = true)
    public Optional<Payment> findByTransactionKey(String transactionKey) {
        log.info("== FIND PAYMENT BY TRANSACTION ID {} ==", transactionKey);
        return jdbcClient.sql(
                "SELECT * FROM Payment p WHERE transactionKey = ?")
            .param(transactionKey)
            .query(Payment.class)
            .optional();
    }

    @Transactional
    public void update(Payment done) {
        log.info("== UPDATE PAYMENT ==");
        String query = "UPDATE Payment SET status = ?, modDate = ? WHERE id = ?";

        jdbcClient.sql(query)
            .params(List.of(done.status().name(), done.modDate(), done.id()))
            .update();
    }
}

File: sinker/src/main/kotlin/com/quid/sinker/pay/gateway/repository/PaymentRepository.kt
package com.quid.sinker.pay.gateway.repository

import com.quid.sinker.pay.domain.Payment
import org.springframework.jdbc.core.simple.JdbcClient
import org.springframework.stereotype.Repository
import org.springframework.transaction.annotation.Transactional

@Repository
class PaymentRepository(
    private val jdbcClient: JdbcClient
) {

    @Transactional
    fun persist(payment: Payment) {
        val query = """
            INSERT INTO payment (id, transaction_key, reference_key, status, amount, currency, reg_date, mod_date)
            VALUES (:id, :transaction_key, :reference_key, :status, :amount, :currency, :reg_date, :mod_date)
            ON CONFLICT (id) DO UPDATE SET 
            transaction_key = EXCLUDED.transaction_key,
            reference_key = EXCLUDED.reference_key,
            status = EXCLUDED.status, 
            amount = EXCLUDED.amount, 
            currency = EXCLUDED.currency,
            reg_date = EXCLUDED.reg_date,
            mod_date = EXCLUDED.mod_date
        """.trimIndent()

        jdbcClient.sql(query)
            .params(
                mapOf(
                    "id" to payment.id,
                    "transaction_key" to payment.transactionKey,
                    "reference_key" to payment.referenceKey,
                    "status" to payment.status.name,
                    "amount" to payment.amount,
                    "currency" to payment.currency,
                    "reg_date" to payment.regDate,
                    "mod_date" to payment.modDate
                )
            ).update()
    }
}